File: Intelligent-Audio-TEST/backend/schemas/audio.py
from __future__ import annotations

from typing import Any, Dict, List, Optional, Union
from pydantic import Field

from backend.schemas.base import APIModel
from backend.schemas.common import PaginatedData
from backend.schemas.testcase import AlgorithmParamItem


class TagListData(APIModel):
    items: List[str] = Field(..., alias='items', validation_alias='items')
    total: int = Field(..., alias='total', validation_alias='total')


class AudioItem(APIModel):
    id: int = Field(..., alias='id', validation_alias='id')
    name: str = Field(..., alias='name', validation_alias='name')
    original_filename: Optional[str] = Field(None, alias='originalFilename', validation_alias='originalFilename')
    file_path: Optional[str] = Field(None, alias='filePath', validation_alias='filePath')
    duration: Optional[float] = Field(None, alias='duration', validation_alias='duration')
    size: Optional[int] = Field(None, alias='size', validation_alias='size')
    sample_rate: Optional[float] = Field(None, alias='sampleRate', validation_alias='sampleRate')
    channels: Optional[int] = Field(None, alias='channels', validation_alias='channels')
    bitrate: Optional[int] = Field(None, alias='bitrate', validation_alias='bitrate')
    format: Optional[str] = Field(None, alias='format', validation_alias='format')
    audio_type: Optional[str] = Field(None, alias='audioType', validation_alias='audioType')
    asr_text: Optional[str] = Field(None, alias='asrText', validation_alias='asrText')
    description: Optional[str] = Field(None, alias='description', validation_alias='description')
    source_language: Optional[str] = Field(None, alias='sourceLanguage', validation_alias='sourceLanguage')
    tags: List[str] = Field(default_factory=list, alias='tags', validation_alias='tags')
    annotations: List[Dict] = Field(default_factory=list, alias='annotations', validation_alias='annotations')
    created_at: Optional[str] = Field(None, alias='createdAt', validation_alias='createdAt')
    updated_at: Optional[str] = Field(None, alias='updatedAt', validation_alias='updatedAt')


class AudioListStats(APIModel):
    total_files: int = Field(..., alias='totalFiles', validation_alias='totalFiles')
    total_size: str = Field(..., alias='totalSize', validation_alias='totalSize')
    total_duration: str = Field(..., alias='totalDuration', validation_alias='totalDuration')
    today_uploads: int = Field(..., alias='todayUploads', validation_alias='todayUploads')


class AudioListData(PaginatedData[AudioItem]):
    stats: AudioListStats = Field(..., alias='stats', validation_alias='stats')


class AudioIdsData(APIModel):
    ids: List[int] = Field(..., alias='ids', validation_alias='ids')
    total: int = Field(..., alias='total', validation_alias='total')


class InitUploadTaskRequest(APIModel):
    pass


class RegisterUploadFileRequest(APIModel):
    task_id: str = Field(..., alias='taskId', validation_alias='taskId')
    files: Optional[List[Dict]] = Field(default_factory=list, alias='files', validation_alias='files')


class AudioAlgorithmRelationItem(APIModel):
    algorithm_type: str = Field(..., alias='algorithmType', validation_alias='algorithmType')
    is_primary: bool = Field(default=False, alias='isPrimary', validation_alias='isPrimary')
    weight: float = Field(default=1.0)
    params: Optional[Dict] = Field(default=None)


class TestCaseUploadConfig(APIModel):
    """上传时携带的测试用例配置，支持多轮 rounds 架构。

    字段全部 optional，不传时由 _create_test_case_from_audio 降级为平面 config。
    """
    rounds: Optional[List[Dict[str, Any]]] = Field(None, alias='rounds', validation_alias='rounds')
    # dimensions 接受 dict 或 list，由 _create_test_case_from_audio 统一处理
    dimensions: Optional[Any] = Field(None, alias='dimensions', validation_alias='dimensions')
    group_name: Optional[str] = Field(None, alias='groupName', validation_alias='groupName')
    inherit_tags: Optional[bool] = Field(True, alias='inheritTags', validation_alias='inheritTags')
    # 算法参数：接受 list（标准 [{field_code, field_value}]）或 dict（{field_code: field_value}），由 controller 归一化
    algorithm_params: Optional[Any] = Field(None, alias='algorithmParams', validation_alias='algorithmParams')


class MergeChunksRequest(APIModel):
    file_id: str = Field(..., alias='fileId', validation_alias='fileId')
    task_id: str = Field(..., alias='taskId', validation_alias='taskId')
    create_test_case: Optional[bool] = Field(False, alias='createTestCase', validation_alias='createTestCase')
    test_types: Optional[List[str]] = Field(default_factory=lambda: ['api'], alias='testTypes', validation_alias='testTypes')
    # dimensions 接受 dict 或 list，由 _create_test_case_from_audio 统一处理
    dimensions: Optional[Any] = Field(default_factory=dict, alias='dimensions', validation_alias='dimensions')
    default_playback_device_id: Optional[int] = Field(None, alias='defaultPlaybackDeviceId', validation_alias='defaultPlaybackDeviceId')
    default_spl: Optional[float] = Field(65.0, alias='defaultSpl', validation_alias='defaultSpl')
    noise_spl: Optional[float] = Field(60.0, alias='noiseSpl', validation_alias='noiseSpl')
    noise_audio_id: Optional[int] = Field(None, alias='noiseAudioId', validation_alias='noiseAudioId')
    test_case_group_name: Optional[str] = Field(None, alias='testCaseGroupName', validation_alias='testCaseGroupName')
    algorithm_type: Optional[str] = Field(None, alias='algorithmType', validation_alias='algorithmType')
    # algorithm_params 接受 list 或 dict，由 controller 归一化
    algorithm_params: Optional[Any] = Field(None, alias='algorithmParams', validation_alias='algorithmParams')
    algorithm_relations: Optional[List[AudioAlgorithmRelationItem]] = Field(None, alias='algorithmRelations', validation_alias='algorithmRelations')
    description: Optional[str] = Field('', alias='description', validation_alias='description')
    tags: Optional[List[str]] = Field(default_factory=list, alias='tags', validation_alias='tags')
    prompt_device_id: Optional[int] = Field(None, alias='promptDeviceId', validation_alias='promptDeviceId')
    prompt_source_language: Optional[str] = Field(None, alias='promptSourceLanguage', validation_alias='promptSourceLanguage')
    prompt_target_language: Optional[str] = Field(None, alias='promptTargetLanguage', validation_alias='promptTargetLanguage')
    prompt_translation_direction: Optional[str] = Field(None, alias='promptTranslationDirection', validation_alias='promptTranslationDirection')
    prompt_algorithm_type: Optional[str] = Field(None, alias='promptAlgorithmType', validation_alias='promptAlgorithmType')
    annotations: Optional[List[Dict]] = Field(default_factory=list, alias='annotations', validation_alias='annotations')
    audio_type: Optional[str] = Field('dry', alias='audioType', validation_alias='audioType')
    asr_text: Optional[str] = Field('', alias='asrText', validation_alias='asrText')
    playback_device_id: Optional[int] = Field(None, alias='playbackDeviceId', validation_alias='playbackDeviceId')
    group_name_type: Optional[str] = Field('root', alias='groupNameType', validation_alias='groupNameType')
    custom_group_name: Optional[str] = Field('', alias='customGroupName', validation_alias='customGroupName')
    inherit_tags: Optional[bool] = Field(True, alias='inheritTags', validation_alias='inheritTags')
    # 多轮上传配置：前端解析文件夹后构建的完整 rounds 配置
    test_case_config: Optional[TestCaseUploadConfig] = Field(None, alias='testCaseConfig', validation_alias='testCaseConfig')

    def get_algorithm_params_dict(self) -> Optional[List[Dict[str, Any]]]:
        """归一化算法参数为 [{field_code, field_value}] 列表格式。

        接受两种输入：
        - list: [{field_code, field_value}, ...] 或 [AlgorithmParamItem, ...]
        - dict: {field_code: field_value, ...}
        """
        if not self.algorithm_params:
            return None
        ap = self.algorithm_params
        # dict 格式：{field_code: field_value}
        if isinstance(ap, dict):
            return [{'field_code': k, 'field_value': v} for k, v in ap.items()]
        # list 格式
        if isinstance(ap, list):
            result = []
            for item in ap:
                if hasattr(item, 'model_dump'):
                    result.append(item.model_dump())
                elif isinstance(item, dict):
                    # 兼容 {field_code, field_value} 和 {fieldCode, fieldValue}
                    fc = item.get('field_code') or item.get('fieldCode')
                    fv = item.get('field_value', item.get('fieldValue'))
                    if fc is not None:
                        result.append({'field_code': fc, 'field_value': fv})
                    else:
                        result.append(item)
            return result if result else None
        return None


class URLImportRequest(APIModel):
    url: str = Field(..., alias='url', validation_alias='url')
    relative_path: Optional[str] = Field(None, alias='relativePath', validation_alias='relativePath')
    audio_type: Optional[str] = Field('dry', alias='audioType', validation_alias='audioType')


class ConvertFormatRequest(APIModel):
    format: str = Field(..., alias='format', validation_alias='format')


class BatchActionRequest(APIModel):
    audio_ids: List[int] = Field(..., alias='audioIds', validation_alias='audioIds')
    action: str = Field(..., alias='action', validation_alias='action')
    tags: Optional[List[str]] = Field(default_factory=list, alias='tags', validation_alias='tags')


class BatchPlaybackRequest(APIModel):
    playback_device_id: Optional[int] = Field(None, alias='playbackDeviceId', validation_alias='playbackDeviceId')
    playback_device_ids: Optional[List[int]] = Field(None, alias='playbackDeviceIds', validation_alias='playbackDeviceIds')
    device_unique_ids: Optional[List[str]] = Field(None, alias='deviceUniqueIds', validation_alias='deviceUniqueIds')
    spl: Optional[float] = Field(None, alias='spl', validation_alias='spl')
    offset: Optional[float] = Field(0, alias='offset', validation_alias='offset')


class UpdateMetadataRequest(APIModel):
    name: Optional[str] = Field(None, alias='name', validation_alias='name')
    audio_type: Optional[str] = Field(None, alias='audioType', validation_alias='audioType')
    asr_text: Optional[str] = Field(None, alias='asrText', validation_alias='asrText')
    description: Optional[str] = Field(None, alias='description', validation_alias='description')
    source_language: Optional[str] = Field(None, alias='sourceLanguage', validation_alias='sourceLanguage')
    tags: Optional[str] = Field(None, alias='tags', validation_alias='tags')
    annotations: Optional[List[Dict]] = Field(default_factory=list, alias='annotations', validation_alias='annotations')


class AudioItemWithAlgorithms(AudioItem):
    algorithms: List[AudioAlgorithmRelationItem] = Field(default_factory=list, alias='algorithms', validation_alias='algorithms')


class UpdateAudioAlgorithmsRequest(APIModel):
    algorithms: List[AudioAlgorithmRelationItem] = Field(..., alias='algorithms', validation_alias='algorithms')


class BatchUpdateAudioAlgorithmsRequest(APIModel):
    audio_ids: List[int] = Field(..., alias='audioIds', validation_alias='audioIds')
    algorithms: List[AudioAlgorithmRelationItem] = Field(..., alias='algorithms', validation_alias='algorithms')

File: Intelligent-Audio-TEST/frontend/src/domain/model/audio.ts
/**
 * Audio（音频资产）领域模型 —— camelCase
 * 对应后端 api_gateway/schemas/audio.py::AudioItem / AudioListData /
 * AudioListStats / AudioIdsData / TagListData / AudioAlgorithmRelationItem
 *
 * 枚举值保持后端原值（如 'dry' / 'noise' / 'prompt'），只转字段名。
 */
import { UploadStatus, TestType, RoundMode, type RoundModeType } from '@/domain/enums'

export type AudioAssetType = 'dry' | 'noise' | 'prompt' | 'mixed' | (string & {})

/** 音频标注条目（后端 annotations: List[Dict]，结构保持宽松但不开放索引签名） */
export interface AudioAnnotation {
  code?: string
  format?: string
  name?: string
  data?: unknown
  sourceLanguage?: string
  targetLanguage?: string
  [key: string]: unknown
}

/** 音频翻译条目（前端本地结构，随上传标注携带） */
export interface AudioTranslation {
  text: string
  direction: string
}

/** 音频文件完整模型（对应 AudioItem） */
export interface AudioInfo {
  id: string | number
  name: string
  originalFilename?: string
  filePath?: string
  size?: number
  duration?: number
  format?: string
  sampleRate?: number
  channels?: number
  bitrate?: number
  /** 音频类型（后端 audio_type 原值） */
  audioType?: AudioAssetType
  asrText?: string
  description?: string
  sourceLanguage?: string
  tags: string[]
  annotations: AudioAnnotation[]
  createdAt?: string
  updatedAt?: string
  /** 前端本地兼容别名（历史消费方依赖，W2 清理） */
  filename?: string
  filepath?: string
  /** 历史本地字段：音频类型（与 audioType 同值，历史消费方使用） */
  type?: AudioAssetType
  translations?: AudioTranslation[]
}

export type Audio = AudioInfo

/** 音频列表统计（对应 AudioListStats） */
export interface AudioListStats {
  totalFiles: number
  /** 人类可读的总大小，如 "1.2 GB" */
  totalSize: string
  /** 人类可读的总时长，如 "03:25:11" */
  totalDuration: string
  todayUploads: number
}

/** 上传中的文件（前端本地状态，非后端契约） */
export interface AudioUploadFile {
  file: File
  id: string
  fileId: string
  name: string
  size: number
  progress: number
  status: typeof UploadStatus[keyof typeof UploadStatus]
  error?: string
  md5?: string
  uploadedSize?: number
  totalChunks?: number
  chunkSize?: number
  uploadedChunks?: number[]
  audioId?: string | number
  folderGroupName?: string
  /** 前端本地兼容别名（历史消费方使用 errorMessage 展示上传失败原因，与 error 同值，W 清理） */
  errorMessage?: string
  /** 最子级文件夹名（无文件夹结构时为去扩展名的文件名），用于按分组独立创建测试用例 */
  groupKey?: string
  asrText?: string
  translations?: AudioTranslation[]
  tags?: string[]
  annotations?: AudioAnnotation[]
}

export interface AudioUploadTask {
  id: string
  files: AudioUploadFile[]
  /** 上传选项（camelCase key，Domain 定义见本文件 AudioUploadOptions） */
  options: AudioUploadOptions
  progress: number
  status: typeof UploadStatus[keyof typeof UploadStatus]
  totalFiles?: number
  completedFiles?: number
  failedFiles?: number
  totalSize?: number
  uploadedSize?: number
  startTime?: string
  endTime?: string
}

/** 音频-算法关联（对应 AudioAlgorithmRelationItem） */
export interface AudioAlgorithmRelation {
  algorithmType: string
  isPrimary: boolean
  weight: number
  params?: Record<string, unknown>
}

/** 标签列表（对应 TagListData：items + total 包装） */
export interface AudioTagList {
  items: string[]
  total: number
}

/** 音频 ID 列表查询结果（对应 AudioIdsData） */
export interface AudioIdsResult {
  ids: number[]
  total: number
}

// ===== 上传相关（前端本地状态，camelCase） =====

/** 上传选项（前端本地状态：camelCase key；发往后端时由 uploadProcess 转换） */
export interface AudioUploadOptions {
  audioType: 'dry' | 'noise' | 'prompt' | 'mixed'
  createTestCase: boolean
  tags: string[]
  description?: string
  testTypes: (typeof TestType[keyof typeof TestType])[]
  /** @deprecated 已移到 CaseForm 的 RoundConfigEditor 里配置 */
  playbackDeviceId?: string | number | null
  /** @deprecated 已移到 CaseForm 的 RoundConfigEditor 里配置 */
  spl?: number
  /** @deprecated 已移到 CaseForm 的 RoundConfigEditor 里配置 */
  noiseAudioId?: string | number | null
  /** @deprecated 已移到 CaseForm 的 RoundConfigEditor 里配置 */
  noiseSpl?: number
  inheritTags?: boolean
  /** 评估维度数组，每条可带 testType 标记属于 api/e2e（由 DimensionConfigData 展开而来） */
  dimensions?: SelectedEvaluationDimension[]
  /** API 维度设置（DimensionConfigPanel 结构：统一/指定轮次/逐轮 + 多轮整体评估维度） */
  apiDimensionConfig?: DimensionConfigData
  /** E2E 维度设置（DimensionConfigPanel 结构：统一/指定轮次/逐轮 + 多轮整体评估维度） */
  e2eDimensionConfig?: DimensionConfigData
  algorithmType?: string
  algorithmRelations?: Array<{
    algorithmType: string
    isPrimary: boolean
    weight: number
    params?: Record<string, any>
  }>
  algorithmParams?: any[]
  promptDeviceId?: string | number | null
  promptSourceLanguage?: string
  promptTargetLanguage?: string
  promptTranslationDirection?: string
  promptAlgorithmType?: string
  groupNameType?: 'root' | 'folder' | 'custom'
  customGroupName?: string
}

/** 评估维度选择项（camelCase 域模型；上行经 Infrastructure toEvaluationDimensionDto 转为后端协议原文 test_type/round_scope） */
export interface SelectedEvaluationDimension {
  id: number | string
  name: string
  weight?: number
  threshold?: number
  /** 标记该维度属于哪种测试类型，'api' / 'e2e'，未标记则通用 */
  testType?: typeof TestType[keyof typeof TestType]
  /** 维度使用范围：'single' = 每轮独立评估，'multi' = 多轮聚合评估。默认 'single' */
  roundScope?: 'single' | 'multi'
  /** 指定生效轮次（specific 模式展开后携带）；未指定则适用所有轮次。-1 表示最后一轮 */
  roundNumber?: number
}

/** 维度配置条目（维度设置面板的输出单元，可带权重/阈值） */
export interface DimensionItem {
  id: string | number
  name: string
  weight?: number
  threshold?: number
}

/** 维度配置数据（上传选项中的维度设置面板结构，对齐 DimensionConfigPanel 输出） */
export interface DimensionConfigData {
  /** 统一模式（all）/ 指定轮次（specific）共用的维度列表 */
  dimensions: DimensionItem[]
  /** 轮次模式：all = 所有轮次统一，specific = 指定轮次，per_round = 逐轮设置 */
  roundMode: RoundModeType
  /** specific 模式下选中的轮次序号列表（-1 = 最后一轮） */
  roundNumbers: number[]
  /** per_round 模式下逐轮维度（key 为轮次序号，-1 = 最后一轮） */
  roundDimensions?: Record<number, DimensionItem[]>
  /** 多轮整体评估维度（roundScope = 'multi'） */
  multiDimensions?: DimensionItem[]
}

/** 评估维度配置 */
export interface EvaluationDimensionsConfig {
  dimensions: SelectedEvaluationDimension[]
}

/** 音频统计摘要（前端展示用） */
export interface AudioStats {
  total: number
  dry: number
  noise: number
  prompt: number
  mixed: number
  totalFiles: number
  totalSize: string
  totalDuration: string
  todayUploads: number
}

/** 音频查询参数（前端 Domain；adapter 负责转 snake_case） */
export interface AudioQueryParams {
  page: number
  perPage: number
  search?: string
  keyword?: string
  type?: string
  /** 标签过滤参数：普通标签为字符串，带模式的标签为 { name, mode } 对象 */
  tags?: (string | { name: string; mode: string })[]
  sortBy?: string
  order?: 'asc' | 'desc'
  audioType?: string
  format?: string
  sampleRate?: string
  duration?: string
  direction?: string
}

/** 文件夹树节点（前端通用视图模型：服务端目录树 / 客户端构建树共用） */
export interface FolderNode {
  name: string
  path?: string
  count?: number
  fileCount?: number
  hasChildren?: boolean
  files: unknown[]
  folders: FolderNode[]
}
